### PR TITLE
Suggestions

### PR DESCRIPTION
**Pull request by Primebrook**
Quite a big PR:

1. Made some changes to `.gitignore`, these were mostly aesthetic changes (e.g. specifying a whole folder rather than individual files).

2. Renamed some files and reorganized some directories. There's three main directories: 

    - `scripts/` - holds preprocessing and training scripts which save outputs to either the `models/` or `data/` directories. 
    - `models/`  - holds the trained price prediction model as a pkl file, this can be unpickled and used to predict on a fresh set of input data points, it also holds the model metrics which contains important info on the model. 
    - `data/` - contains scraped data and cleaned data i.e. data after preprocessing.

4. Added the training script, with a trained model and model metrics. This is the biggest addition. It's an XGBoost Regression model, which uses the hyperopt library along with K Fold cross validation (default `n_folds = 10`) to optimize the hyperparameters (hyperopt uses some bayesian magic to search the parameter space more efficiently and effectively than say sklearn's RandomSearchCV or GridSearchCV.  The model uses the root mean squared error as th loss function. The target variable of the model is the log of the price, so well using for predictions this transformation will have to be reversed to get the actual price.

5. Added a `requirements.txt` file (childish that it didn't exist already tbh 🙄)

6. Added types annotations to the `data_cleaning.py` file (added type annotation on everything I added) and made some small other edits/refactoring.

### CODE REVIEW

File: data_preparation/data_cleaning.py [hunk 1]
```diff
@@ -215,9 +207,17 @@ def fill_columns_with_missing_values(df):
             df[column] = df.groupby(['make', 'model', 'trim', 'manufactured_year'])[column].transform(lambda x: x.fillna(x.mode()[0] if not x.mode().empty else np.nan))
             df[column] = df.groupby(['make', 'model', 'trim'])[column].transform(lambda x: x.fillna(x.mode()[0] if not x.mode().empty else np.nan))
     
-    return df    
+    return df
+
+def train_no_of_owners_model(df: pd.DataFrame) -> np.ndarray:
```
Comment (ezeahunanya) [inline data_preparation/data_cleaning.py]: Why have you added this function?

Comment (ezeahunanya) [reply data_preparation/data_cleaning.py]: Also whats the difference between me adding this comment and starting a review

File: data_preparation/data_cleaning.py [hunk 2]
```diff
@@ -130,30 +128,24 @@ def clean_round1(df):
...
-    df['transmission'] = df.transmission.map(lambda x: x.strip().lower().replace(' ', '_'), na_action='ignore').astype('category')
-    df['fuel_type'] = df.fuel_type.map(lambda x: x.strip().lower().replace(' ', '_'), na_action='ignore').astype('category')
-    df['insurance_group'] = df.insurance_group.map(lambda x: x.strip().lower().replace(' ', '_'), na_action='ignore').astype('category')
-    df['seller_segment'] = df.seller_segment.map(lambda x: x.strip().lower().replace(' ', '_'), na_action='ignore').astype('category')
-    df['region'] = df.region.map(lambda x: x.strip().lower().replace(' ', '_'), na_action='ignore').astype('category')
-    df['county'] = df.county.map(lambda x: x.strip().lower().replace(' ', '_'), na_action='ignore').astype('category')
-    df['town'] = df.town.map(lambda x: x.strip().lower().replace(' ', '_'), na_action='ignore').astype('category')
+    cols = ['make','model','trim','body_type','transmission','fuel_type','insurance_group','seller_segment','region','county','town']
+    sep = defaultdict(lambda:'_', {'make':'-', 'model':'-', 'trim':'-'})
```
Comment (ezeahunanya) [inline data_preparation/data_cleaning.py:128]: You know how i used _ and - i don't think there was an actual reason other than preferring how it looked in make, model and trim columns. ie mercedes-benz vs mercedes_benz. I think you could change it all to _ if that makes the code simpler

File: data_preparation/data_cleaning.py [hunk 3]
```diff
@@ -278,3 +278,10 @@ def fix_int_datatypes(df):
                         'co2_emissions': 'int', 'total_reviews': 'int'})
                         
     return df
+
+
+if __name__ == '__main__':
+    df = pd.read_csv('/Users/brook/downloads/data.csv')
```
Comment (ezeahunanya) [inline data_preparation/data_cleaning.py]: I think this line is too specific to your file structure, could it be written so i could run it with the data.csv file in the top level folder?

File: scripts/preprocess_price_prediction_model.py [hunk 4]
```diff
@@ -49,21 +42,28 @@ def drop_columns(df):
...
+    '''
+    Combine CO2 columns into one.
+    '''
+    co2_emissions = df[(df.co2_emissions.isnull()) & (~df.co2_emission.isnull())].co2_emission
+    
+    for index, value in co2_emissions.items():
+        df.loc[index, 'co2_emissions'] = value
+
+    return df.drop('co2_emission', axis=1)
+
```
Comment (ezeahunanya) [inline scripts/preprocess_price_prediction_model.py:57]: Was this copied from the cleaning file because you couldn't import it?

File: scripts/train_price_prediction_model.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+
+if __name__ == '__main__':
+    training_data = pd.read_csv(PROJ_DIR+'/data/training_data.csv')
+    X = training_data[features]
+    y = training_data['price'].apply(lambda x: np.log(1+x)) # Changing target variable to log of price cause price is massively right skewed
+    model, model_details = main(X, y)
+    pprint(model_details, sort_dicts=False)
+    save_as_pickle(obj=model, filepath=PROJ_DIR+'/models/price_prediction_model.pkl')
+    save_as_json(obj=model_details, filepath=PROJ_DIR+'/models/price_prediction_metrics.json')
```
Comment (ezeahunanya) [inline scripts/train_price_prediction_model.py:110]: ayy big man teach meee how to build models like this

File: data_preparation/data_cleaning.py [hunk 6]
```diff
@@ -1,41 +1,37 @@
...
-#                   ORDER BY date_scraped ASC, time_scraped ASC''')
+    cnx = mysql.connector.connect(**mysql_details)
+    cursor = cnx.cursor(dictionary=True)
+
+    cursor.execute("USE {}".format(DB_NAME))
+    cursor.execute('''SELECT * 
+                  FROM vehicle_features as vf
+                  LEFT JOIN sellers as s
+                  ON vf.seller_id=s.seller_id
+                  ORDER BY date_scraped ASC, time_scraped ASC''')
```
Comment (ezeahunanya) [inline data_preparation/data_cleaning.py]: tbf you could leave this commented out permanently and we could just work from the csv file instead

File: readme.md [hunk 7]
```diff
@@ -27,6 +27,15 @@ from the command line. Then run the following command:
...
+To create the training data for the price prediction model run: 
+```
+python scripts/preprocess_price_prediction_model.py
+```
+
+This will output training data to `data/training_data.csv`. To train the model run 
+```
+python scripts/train_price_prediction_model.py
+``` 
+This will output the trained model and the model metrics (optimized hyperparameters, evaluation metrics, feature importances) and save to `models/`.
```
Comment (ezeahunanya) [inline readme.md]: my g

File: data_preparation/data_cleaning.py (hunk 1 above)
Comment (Primebrook) [reply data_preparation/data_cleaning.py]: I removed this change in a later commit. If you go on the Files Changed tab and review each file then it'll show you the final version of changes I made. If you click on a specific commit and review one of those changes (as you've done here) then you're not necessarily looking at the latest changes on that file i.e. I could have reverted that change in a later commit (which is the case here). I think you've done this on one of you other comments, I'll just say `SEE FINAL FILE VERSION` in reply where you've done it.

And what you're doing now (writing comments) is the reviewing process. You can make comments, propose further changes etc and then when you're finally happy with everything you can add your approval to this PR, which is you saying that you're okay with it being merged into master. Then it's up to me to merge it into master (you call you master branch main for this repo actually). Cause I'm a collaborator on this repo I can merge into master without approval also. But making a PR and reviewing is good practice especially for non-trivial changes.

File: data_preparation/data_cleaning.py (hunk 3 above)
Comment (Primebrook) [reply data_preparation/data_cleaning.py]: `SEE FINAL FILE VERSION`

File: scripts/preprocess_price_prediction_model.py (hunk 4 above)
Comment (Primebrook) [reply scripts/preprocess_price_prediction_model.py:57]: I did it cause it was only two functions and I the purpose of the preprocessing script is to have all the preprocessing/cleaning functions in them. I wanted it to be self contained. There's still more changes to make on the preprocessing script I think. I won't have to make changes to the `data_cleaning.py` file from now on.

File: scripts/train_price_prediction_model.py (hunk 5 above)
Comment (Primebrook) [reply scripts/train_price_prediction_model.py:110]: It's a light oneee 👀

File: data_preparation/data_cleaning.py (hunk 6 above)
Comment (Primebrook) [reply data_preparation/data_cleaning.py]: Hmmm okay, I guess we can setup the db stuff later on if you want it. Though the file sizes we're dealing with, don't think it's strictly needed, but still could be good practice setting it up.

File: data_preparation/data_cleaning.py (hunk 2 above)
Comment (Primebrook) [reply data_preparation/data_cleaning.py:128]: Hmm okay, I'll look at it again, I thought I remembered there being a place where you relied on the dash as opposed to the underscore, but could be wrong. Also you don't actually use this function in the preprocessing script anyway?

File: data_preparation/data_cleaning.py (hunk 1 above)
Comment (ezeahunanya) [reply data_preparation/data_cleaning.py]: Nice

File: scripts/preprocess_price_prediction_model.py (hunk 4 above)
Comment (ezeahunanya) [reply scripts/preprocess_price_prediction_model.py:57]: ahh i see, makes sense

File: scripts/train_price_prediction_model.py (hunk 5 above)
Comment (ezeahunanya) [reply scripts/train_price_prediction_model.py:110]: Nah bmt though

File: data_preparation/data_cleaning.py (hunk 6 above)
Comment (ezeahunanya) [reply data_preparation/data_cleaning.py]: I mean i don't think we'll need the database anyways.

Comment (Primebrook) [reply data_preparation/data_cleaning.py]: okay calm

File: scripts/train_price_prediction_model.py (hunk 5 above)
Comment (Primebrook) [reply scripts/train_price_prediction_model.py:110]: yeah that's what this collab process is about

File: data_preparation/data_cleaning.py (hunk 2 above)
Comment (ezeahunanya) [reply data_preparation/data_cleaning.py:128]: Yeah you're right it was just for the initial eda but i don't use it for the main model pre processing
